oqamaint/openqabot.py

```diff
@@ -0,0 +1,588 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import print_function
+
+from collections import namedtuple
+from datetime import date
+import md5
+from pprint import pformat
+import re
+
+
+import requests
+import osc.core
+
+import ReviewBot
+from osclib.comments import CommentAPI
+
+from suse import SUSEUpdate
+
+try:
+    from xml.etree import cElementTree as ET
+except ImportError:
+    from xml.etree import ElementTree as ET
+
+
+try:
+    import simplejson as json
+except ImportError:
+    import json
+
+QA_UNKNOWN = 0
+QA_INPROGRESS = 1
+QA_FAILED = 2
+QA_PASSED = 3
+
+Package = namedtuple('Package', ('name', 'version', 'release'))
+pkgname_re = re.compile(r'(?P<name>.+)-(?P<version>[^-]+)-(?P<release>[^-]+)\.(?P<arch>[^.]+)\.rpm')
+comment_marker_re = re.compile(
+    r'<!-- openqa state=(?P<state>done|seen)(?: result=(?P<result>accepted|declined|none))?(?: revision=(?P<revision>\d+))? -->')
```

Comment:
As noted on @coolo's submission, I did a lot of work on comment code in `ReviewBot` and `osclib/comments.py` lets use that instead of duplicating it in openqa bot. See `leaper.py` or `repo_checker.py` for fairly extensive use of the comment stuff. Also note the `ReviewBot` also has the ability to automatically collect comment content from a logger like leaper uses, but all the replacing old comment code should remove the need for this regex and related code.

Reply:
No, unfortunately, most important part of this regex is the **revision** on which depends the whole functionality of bot

Reply:
The code supports a dict of key values. "revision=<version>" is a key and value...how will it not work?

Reply:
Note that `repo_checker` even generates a hash which is essentially the same as the revision used here.

https://github.com/openSUSE/osc-plugin-factory/blob/3bbeacd75f0a99b725a92c2689c91efa621657df/repo_checker.py#L74-L86

Reply:
All that is needed is `ReviewBot.comment_write()` to provide optional additional info and pass it along to `CommentAPI.comment_write()`.

Currently, it just builds to dict from the only two keys that bots used, but that's a restriction present in `ReviewBot` only. Considering how much duplicate code this eliminates it's worth the tiny refactoring.

https://github.com/openSUSE/osc-plugin-factory/blob/3bbeacd75f0a99b725a92c2689c91efa621657df/ReviewBot.py#L456

Reply:
Even if `ReviewBot` was not refactored the `CommentAPI` methods should be used instead of re-inventing them as well, but really I've seen improvements from `ReviewBot` copied here and the code short of the `revision` does the exact same thing and trivial to implement. Anytime one sees that it's a case for refactoring not copying.

Reply:
revision in this case is totallu different  thing .. and calculating has of binaries in osclib is prettty stupid .. OBS repodata.xml has revision , simply read it

Reply:
What are you talking about. I embedded the code related to hash it has nothing to do with the one OBS provides. The only feature you want is a single extra key=value pair in addition to state and result. The `CommentAPI` functions were enhanced to provide just that. The `ReviewBot` code simply needs to be changed from only providing `state` and `result` to either just an `info` dict or adding an `additional_info` parameter.

That is the only thing missing. In fact the other methods used to insert the marker and find existing comments can be used directly from `CommentAPI` just fine. The `ReviewBot` version is just the highest level wrapper that does everything and unfortunately limits this one, teany, tiny feature for no reason other than it was never needed before.

Why you are assuming so much instead of just reading when I spent a fair amount of time reviewing this to try and improve it is utterly beyond me. I made the effort to embed the relevant code bits and you clearly didn't even skim ready them (all 10 lines of them).

Reply:
Tada! https://github.com/openSUSE/osc-plugin-factory/pull/1154/commits/b886949e836efd45367bfa748dec4530fc90bc14 I am still sweating from the effort.

Reply:
I'll plan to rewrite it to comment api. "revision" used in QAM requests/incidents is different than hash you use for differentiating comments. Our revision is highest repository revision from channels. If it changes we retrigger needed job. And also it can be exactly read in another internal tooling.

Reply:
The revision in the comment to tell if the already posted results are for same thing as last time? It really does not matter, from comment API it is simply used to match a previous comment. So as long as you can specify `revision` key and the value as whatever you want and have it match it behaves as you need. I never implied the one in `repo-checker` was the same, merely that is is an arbitrary hash used for comment matching...